File: integration-test/src/main/java/com/sequenceiq/it/util/CleanupService.java
package com.sequenceiq.it.util;

import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import com.sequenceiq.cloudbreak.api.model.users.WorkspaceResponse;
import com.sequenceiq.cloudbreak.client.CloudbreakClient;
import com.sequenceiq.it.cloudbreak.CloudbreakUtil;
import com.sequenceiq.it.cloudbreak.WaitResult;
import com.sequenceiq.it.cloudbreak.config.ITProps;

@Component
public class CleanupService {
    private static final Logger LOG = LoggerFactory.getLogger(CleanupService.class);

    private static final int DELETE_SLEEP = 30000;

    private boolean cleanedUp;

    @Value("${integrationtest.cleanup.retryCount}")
    private int cleanUpRetryCount;

    @Inject
    private ITProps itProps;

    public synchronized void deleteTestStacksAndResources(CloudbreakClient cloudbreakClient) {
        if (cleanedUp) {
            return;
        }
        cleanedUp = true;
        cloudbreakClient.workspaceV3Endpoint().getAll()
                .stream()
                .filter(workspace -> workspace.getName().startsWith("its-"))
                .map(WorkspaceResponse::getId)
                .forEach(workspaceId -> cloudbreakClient.stackV3Endpoint().listByWorkspace(workspaceId).stream()
                                .filter(stack -> stack.getName().startsWith("it-"))
                                .forEach(stack -> deleteStackAndWait(cloudbreakClient, workspaceId, stack.getName()))
                );

        cloudbreakClient.blueprintEndpoint()
                .getPrivates()
                .stream()
                .filter(blueprint -> blueprint.getName().startsWith("it-"))
                .forEach(blueprint -> deleteBlueprint(cloudbreakClient, String.valueOf(blueprint.getId())));

        cloudbreakClient.recipeEndpoint()
                .getPrivates()
                .stream()
                .filter(recipe -> recipe.getName().startsWith("it-"))
                .forEach(recipe -> deleteRecipe(cloudbreakClient, recipe.getId()));

        cloudbreakClient.workspaceV3Endpoint().getAll()
                .stream()
                .filter(workspace -> workspace.getName().startsWith("its-"))
                .map(WorkspaceResponse::getId)
                .forEach(workspaceId -> cloudbreakClient.credentialV3Endpoint().listByWorkspace(workspaceId).stream()
                        .filter(c -> "AZURE".equals(c.getCloudPlatform()) ? c.getName().startsWith("its") : c.getName().startsWith("its-"))
                        .forEach(credential -> deleteCredential(cloudbreakClient, workspaceId, credential.getName()))
                );

        cloudbreakClient.rdsConfigEndpoint()
                .getPrivates()
                .stream()
                .filter(rds -> rds.getName().startsWith("it-"))
                .forEach(rds -> deleteRdsConfigs(cloudbreakClient, rds.getId().toString()));

        cloudbreakClient.workspaceV3Endpoint().getAll()
                .stream()
                .filter(workspace -> workspace.getName().startsWith("its-"))
                .forEach(workspace -> deleteWorkspace(cloudbreakClient, workspace.getName()));
    }

    public void deleteWorkspace(CloudbreakClient cloudbreakClient, String name) {
        if (name != null) {
            cloudbreakClient.workspaceV3Endpoint().deleteByName(name);
        }
    }

    public void deleteCredential(CloudbreakClient cloudbreakClient, Long workspaceId, String credentialName) {
        if (credentialName != null && workspaceId != null) {
            cloudbreakClient.credentialV3Endpoint().deleteInWorkspace(workspaceId, credentialName);
        }
    }

    public void deleteBlueprint(CloudbreakClient cloudbreakClient, String blueprintId) {
        if (blueprintId != null) {
            cloudbreakClient.blueprintEndpoint().delete(Long.valueOf(blueprintId));
        }
    }

    public void deleteStackAndWait(CloudbreakClient cloudbreakClient, Long workspaceId, String stackName) {
        for (int i = 0; i < cleanUpRetryCount; i++) {
            if (deleteStack(cloudbreakClient, workspaceId, stackName)) {
                WaitResult waitResult = CloudbreakUtil.waitForStackStatus(cloudbreakClient, workspaceId, stackName, "DELETE_COMPLETED");
                if (waitResult == WaitResult.SUCCESSFUL) {
                    break;
                }
                try {
                    Thread.sleep(DELETE_SLEEP);
                } catch (InterruptedException e) {
                    LOG.warn("interrupted ex", e);
                }
            }
        }
    }

    public boolean deleteStack(CloudbreakClient cloudbreakClient, Long workspaceId, String stackName) {
        boolean result = false;
        if (stackName != null && workspaceId != null) {
            cloudbreakClient.stackV3Endpoint().deleteInWorkspace(workspaceId, stackName, false, false);
            result = true;
        }
        return result;
    }

    public void deleteRecipe(CloudbreakClient cloudbreakClient, Long recipeId) {
        cloudbreakClient.recipeEndpoint().delete(recipeId);
    }

    public void deleteImageCatalog(CloudbreakClient cloudbreakClient, String name) {
        cloudbreakClient.imageCatalogEndpoint().deletePublic(name);
    }

    public void deleteRdsConfigs(CloudbreakClient cloudbreakClient, String rdsConfigId) {
        if (rdsConfigId != null) {
            cloudbreakClient.rdsConfigEndpoint().delete(Long.valueOf(rdsConfigId));
        }
    }
}

File: integration-test/src/main/java/com/sequenceiq/it/util/CleanupService.kt
package com.sequenceiq.it.util

import javax.inject.Inject

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

import com.sequenceiq.cloudbreak.api.endpoint.SecurityGroupEndpoint
import com.sequenceiq.cloudbreak.api.model.BlueprintResponse
import com.sequenceiq.cloudbreak.api.model.CredentialResponse
import com.sequenceiq.cloudbreak.api.model.NetworkJson
import com.sequenceiq.cloudbreak.api.model.RecipeResponse
import com.sequenceiq.cloudbreak.api.model.SecurityGroupJson
import com.sequenceiq.cloudbreak.api.model.StackResponse
import com.sequenceiq.cloudbreak.api.model.TemplateResponse
import com.sequenceiq.cloudbreak.client.CloudbreakClient
import com.sequenceiq.it.cloudbreak.CloudbreakUtil
import com.sequenceiq.it.cloudbreak.WaitResult
import com.sequenceiq.it.cloudbreak.config.ITProps

@Component
class CleanupService {

    private var cleanedUp: Boolean = false

    @Value("${integrationtest.cleanup.retryCount}")
    private val cleanUpRetryCount: Int = 0
    @Inject
    private val itProps: ITProps? = null

    @Synchronized @Throws(Exception::class)
    fun deleteTestStacksAndResources(cloudbreakClient: CloudbreakClient) {
        if (cleanedUp) {
            return
        }
        cleanedUp = true
        val stacks = cloudbreakClient.stackEndpoint().privates
        for (stack in stacks) {
            if (stack.name!!.startsWith("it-")) {
                deleteStackAndWait(cloudbreakClient, stack.id.toString())
            }
        }
        val templates = cloudbreakClient.templateEndpoint().privates
        for (template in templates) {
            if (template.name!!.startsWith("it-")) {
                deleteTemplate(cloudbreakClient, template.id.toString())
            }
        }
        val networks = cloudbreakClient.networkEndpoint().privates
        for (network in networks) {
            if (network.name!!.startsWith("it-")) {
                deleteNetwork(cloudbreakClient, network.id.toString())
            }
        }
        val secgroups = cloudbreakClient.securityGroupEndpoint().privates
        for (secgroup in secgroups) {
            if (secgroup.name!!.startsWith("it-")) {
                deleteSecurityGroup(cloudbreakClient, secgroup.id.toString())
            }
        }
        val blueprints = cloudbreakClient.blueprintEndpoint().privates
        for (blueprint in blueprints) {
            if (blueprint.name!!.startsWith("it-")) {
                deleteBlueprint(cloudbreakClient, blueprint.id.toString())
            }
        }
        val recipes = cloudbreakClient.recipeEndpoint().privates
        for (recipe in recipes) {
            if (recipe.name!!.startsWith("it-")) {
                deleteRecipe(cloudbreakClient, recipe.id)
            }
        }
        val credentials = cloudbreakClient.credentialEndpoint().privates
        for (credential in credentials) {
            if ("AZURE_RM" == credential.cloudPlatform && credential.name!!.startsWith("its") || "AZURE_RM" != credential.cloudPlatform && credential.name!!.startsWith("its-")) {
                deleteCredential(cloudbreakClient, credential.id.toString())
            }
        }
    }

    @Throws(Exception::class)
    fun deleteCredential(cloudbreakClient: CloudbreakClient, credentialId: String?): Boolean {
        var result = false
        if (credentialId != null) {
            cloudbreakClient.credentialEndpoint().delete(java.lang.Long.valueOf(credentialId))
            result = true
        }
        return result
    }

    @Throws(Exception::class)
    fun deleteTemplate(cloudbreakClient: CloudbreakClient, templateId: String?): Boolean {
        var result = false
        if (templateId != null) {
            cloudbreakClient.templateEndpoint().delete(java.lang.Long.valueOf(templateId))
            result = true
        }
        return result
    }

    @Throws(Exception::class)
    fun deleteNetwork(cloudbreakClient: CloudbreakClient, networkId: String?): Boolean {
        var result = false
        if (networkId != null) {
            cloudbreakClient.networkEndpoint().delete(java.lang.Long.valueOf(networkId))
            result = true
        }
        return result
    }

    @Throws(Exception::class)
    fun deleteSecurityGroup(cloudbreakClient: CloudbreakClient, securityGroupId: String?): Boolean {
        var result = false
        if (securityGroupId != null) {
            val securityGroupEndpoint = cloudbreakClient.securityGroupEndpoint()
            val securityGroupJson = securityGroupEndpoint[java.lang.Long.valueOf(securityGroupId)]
            if (securityGroupJson.name != itProps!!.defaultSecurityGroup) {
                securityGroupEndpoint.delete(java.lang.Long.valueOf(securityGroupId))
                result = true
            }
        }
        return result
    }

    @Throws(Exception::class)
    fun deleteBlueprint(cloudbreakClient: CloudbreakClient, blueprintId: String?): Boolean {
        var result = false
        if (blueprintId != null) {
            cloudbreakClient.blueprintEndpoint().delete(java.lang.Long.valueOf(blueprintId))
            result = true
        }
        return result
    }

    @Throws(Exception::class)
    fun deleteStackAndWait(cloudbreakClient: CloudbreakClient, stackId: String): Boolean {
        var deleted = false
        for (i in 0..cleanUpRetryCount - 1) {
            if (deleteStack(cloudbreakClient, stackId)) {
                val waitResult = CloudbreakUtil.waitForStackStatus(cloudbreakClient, stackId, "DELETE_COMPLETED")
                if (waitResult == WaitResult.SUCCESSFUL) {
                    deleted = true
                    break
                }
                try {
                    Thread.sleep(DELETE_SLEEP.toLong())
                } catch (e: InterruptedException) {
                    LOG.warn("interrupted ex", e)
                }

            }
        }
        return deleted
    }

    @Throws(Exception::class)
    fun deleteStack(cloudbreakClient: CloudbreakClient, stackId: String?): Boolean {
        var result = false
        if (stackId != null) {
            cloudbreakClient.stackEndpoint().delete(java.lang.Long.valueOf(stackId), false)
            result = true
        }
        return result
    }

    @Throws(Exception::class)
    fun deleteRecipe(cloudbreakClient: CloudbreakClient, recipeId: Long?): Boolean {
        cloudbreakClient.recipeEndpoint().delete(recipeId)
        return true
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(CleanupService::class.java)
        private val DELETE_SLEEP = 30000
    }
}
